### PR TITLE
[dask.order] Reduce memory pressure for multi array reductions by releasing splitter tasks more eagerly

### PR DESCRIPTION
**Pull request by fjetter**
Related https://github.com/dask/dask/pull/10505
Closes https://github.com/dask/dask/issues/10384
Closes https://github.com/dask/dask/issues/9995

- It removes the ambiguity between dataframe and array computes by normalizing topologies by adding a final reducing node to all graphs (this requires a little more work)
- It adds branched out dependents to the singles list. this is not exactly what it was designed for but this allows us to process those tasks before opening a new branch of computation
- It removes a "change strategic goal" code path. Mostly because it is necessary to make the code work but also because I currently distrust the partition_keys target function

Edit: The above description is a little outdated. See comment farthest below and inline comments for an accurate description

### CODE REVIEW

File: dask/tests/test_order.py [hunk 1]
```diff
@@ -109,22 +109,19 @@ def test_base_of_reduce_preferred(abcde):
...
-         |
-         a3    d1
-        /  \  /
-      b1    c1
-      |     |
-      b2    c2
-            |
-            c3
-
-    Prefer b1 over c1 because it won't stick around waiting for d1 to complete
```
Comment (fjetter) [inline dask/tests/test_order.py:127]: I believe this explanation is actually wrong. I think `c3` should be preferred to run before `b2` because `c3->a1` is the longest critical path. 

I'm not even entirely convinced that `d1` should run before `b2`. I believe the memory pressure is identical and `b2` has a longer way to walk. Anyway, this is how it currently works and I think that's fine

File: dask/tests/test_order.py [hunk 2]
```diff
@@ -301,36 +299,24 @@ def test_run_smaller_sections(abcde):
      / \ /|  | /
     a   c e  cc
 
-    Prefer to run acb first because then we can get that out of the way
```
Comment (fjetter) [inline dask/tests/test_order.py:304]: This is currently favoring finishing `aa` before anything else. I believe this is a matter of finding the appropriate target functions.

Considering that it was xfailed I'm not terribly concerned but I will have a look again.

**Review by fjetter — COMMENTED** [review]
Currently, there are still failing tests

- `test_prefer_short_dependents` requires me to adjust the "finish now" logic to account for the artificial root task
- `test_order_with_equal_dependents` TODO
- `test_run_smaller_sections` see comment

**Comment by fjetter** [comment]
At least on smaller graphs, this is fixing the root task problem mentioned in https://github.com/dask/dask/issues/9995 by ensuring to finish a computation branch before opening a new one
anom-mean-order-raw-colored

The xarray style reduction reported in https://github.com/dask/dask/issues/10384 now looks like 
xarray-reduction-array-colored
the simulated pressure for dataframes and arrays are not identical due to the adjustment to the graph topologies. What's even better, the pressure with this PR is now smaller than the DF pressure before.

I will move to real world testing/benchmarking tomorrow. If that looks good, I will have to do a little refactoring since there is now plenty of dead code I'm effectively skipping due to the root task addition. 🤞

File: dask/order.py [hunk 3]
```diff
@@ -133,6 +134,14 @@ def order(dsk, dependencies=None):
     # tree, we skip processing it normally.
     # See https://github.com/dask/dask/issues/6745
     root_nodes = {k for k, v in dependents.items() if not v}
+    if len(root_nodes) > 1:
+        root = "root-node"
+        dsk[root] = (object(), *root_nodes)
+        dependencies[root] = root_nodes
+        o = order(dsk, dependencies)
+        del o[root]
+        return o
```
Comment (fjetter) [inline dask/order.py:143]: This is by far the most interesting change in this PR since it is effectively disabling a lot of logic. For one, this is disabling the entire `skip_root_node` and also specialized `finish_now` things.

What's even more critical is that this effectively disables many root-node dependent graph metrics (stuff like max_dependencies, min_dependencies) since all nodes now share the same goal

Comment (fjetter) [reply dask/order.py:143]: There is an interesting effect, though. The partition_keys function is depending on `min_dependencies`, or rather the difference between `min_dependencies` and `total_dependencies`. If min_dependencies was removed entirely, this function would become negative, reverting every decision.

**Comment by fjetter** [comment]
Ok, I got through the unit tests and think everything is in order. I got sidetracked a little when looking at the cost functions because the graph_metrics confused me a little. Turns out (obvious in hindsight) they are only accurate for trees and are double counting quite a bit. Oh well, now there are a couple of tests. I also changed the definition slightly in a way that makes more sense to me. I may break this out into a dedicated PR, this is quite unrelated to the important stuff in here

**Comment by mrocklin** [comment]
Cool.  How excited should I be about this?  "Solves all Xarray pain?"  "Solves a couple of problems we've seen recently, but still lots more to do?"

**Comment by fjetter** [comment]
> Cool. How excited should I be about this? "Solves all Xarray pain?" "Solves a couple of problems we've seen recently, but still lots more to do?"

Let's wait for the first round of benchmarks before I answer :)

**Comment by fjetter** [comment]
OK, so benchmark results will be a little delayed. There is some kind of issue with our CI.

From manual testing I can say that the xarray example that's been reported in https://github.com/pangeo-data/distributed-array-examples/issues/2 now runs smoothly with constant memory.
However, if I dial up the size, we are running again into a bad ordering problem. I suspect something with the graph metrics is off.

(Legend below: scaling the time dimension / number of tasks by that factor; The green line is only constant because we're spilling)

image

The up side is that on main, the 1.0 measure was already saturating memory fully... Looks like I have some more work to do

**Comment by fjetter** [comment]
Well, at the very least simulated memory pressure, i.e. the number of tasks a single threaded scheduler would have to hold in memory, is behaving reasonably well. Not constant but no [longer linear](https://github.com/dask/dask/issues/10384#issuecomment-1715841811) and even better than the DF version before.

(This goes up to 100_000, i.e. 20x more than the reported example)

image

**Comment by fjetter** [comment]
I'm pretty sure that what I'm seeing right now is not an ordering problem anymore. The behavior appears to be quite flaky. I assume this there is some kind distributed scheduler state that is impacting the worker placement decision.

For instance, I now managed to get another round of runs through where the _small_ dataset actually misbehaved but I didn't have an issue scaling up to 5 times the data (i.e. total input data is 10x more than the cluster memory)

image

**Comment by fjetter** [comment]
OK, turns out it is an ordering problem after all. Order appears to not be stable.

I patched the scheduler to compute a hash of the order result and compute the simulated pressure and log this. The hash is always different. The pressure is sometimes bad and when it is, the run is horrible.

Just running the same thing in a for loop for a couple of times

image

At least I know what to look for now... kind of

**Comment by fjetter** [comment]
Well... good and bad news. Order seems to be fine (ish?). If I disable optimization, everything seems to be very, very stable. However, once optimization is on, this becomes flaky!
going through the code, once I remove/comment out `optimize_blockwise` everything works fine. With `optimize_blockwise` this is flaky

**Comment by fjetter** [comment]
By best guess is that this is connected to keys somehow. This issue is only flaky if I regenerate the graph, i.e. regenerate the keys. We probably no longer have sufficient structural information to break ties w/out str comp after the fusion.

**Comment by fjetter** [comment]
TLDR I shouldn't have touch the graph_metrics. However, everything is sooo fragile that there is more work to do still

Who was I to think touching the graph_metrics definition was a good idea? Stupid me! The two diverging orderings were indeed triggered by a str comparison but the fundamental bug was introduced by me changing the graph_metrics. I don't know if this will ever be relevant but for my own sanity, I want to document what happened. Interested readers may skip this safely.

I was looking at why these two graphs diverged

raw graph
first

snapshot good ordering
image


snapshot bad ordering

image


The good ordering properly targeted the first reducer and (not shown here) would clean up after itself. That's perfect. The bad ordering regressed into a breadth first search that was releasing roots relatively quickly but wouldn't go deep, such that we cannot truly release anything until rather late.

The divergence happens already at the very fist decision point. The initial root node is random (str compared) because the graph is pretty much symmetrical. It would then decide which one of the dependents to pick by using `dependents_key` which is identical for both after my graph_metrics changes (because with my changes, _coincidentally_ min_heights and total_dependents is equal causing this to zero out https://github.com/dask/dask/blob/2b45b215e13e6d7893853ef33cb4b1796c477a9e/dask/order.py#L224C47-L224C47)
This decision here https://github.com/dask/dask/blob/2b45b215e13e6d7893853ef33cb4b1796c477a9e/dask/order.py#L593-L599

which leads us to these if/elif/else clauses

https://github.com/dask/dask/blob/2b45b215e13e6d7893853ef33cb4b1796c477a9e/dask/order.py#L640-L652

The thing is, these dependents are **not** identical since one of them can already be run while the other has a dependency to load first. Depending on which one of the two is `dep2` the `if not num_needed[dep2]` clause will evaluate differently.
The bad case actually runs in to the `num_needed[dep2] == 0` zero case and puts a "single" relation on the other dependency
https://github.com/dask/dask/blob/2b45b215e13e6d7893853ef33cb4b1796c477a9e/dask/order.py#L645 which forces `order` to work on this "single" _before_ loading any more dependents. Effectively these _singles_ is a deviation from the depth first search (which is sometimes necessary to free memory).
The good path actually picks the `key == key2 and 5 * partition_keys[item] > 22 * key` path (where do these magical numbers come from) https://github.com/fjetter/dask/blob/e71033821ecedab254750bd6f158714ba589acf8/dask/order.py#L650
which does almost the same thing but doesn't put a "single" relationship on the other dep.

Alas, reverting the graph_metrics change ensure that both graphs are now producing identical orderings. The chosen path is actually the bad one... but once we move to larger graphs it appears to work again. This just shows that there is still some work to do...

**Comment by fjetter** [comment]
alright, the above example now works nicely with constant memory. The below shows runs with increasingly more data/tasks. Runtime is increasing linearily and memory is constant. Just the way it's supposed to.

image

**Comment by dcherian** [comment]
❤️

File: dask/tests/test_order.py (hunk 2 above)
Comment (fjetter) [reply dask/tests/test_order.py:304]: This example is a bit ambiguous. There is a "best ordering" for single threaded execution and one for multi threaded execution.

Single threaded

This PR
| Step     | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
| -        | - | -- | -- | -- | - | - | - | - | - |
| T1       | e | c  | d | cc | bb | aa | dd | a | b |
| Pressure | 1 | 2  | 2  | 3  | 4 | 3  | 3  | 4 | 2 |

What the docs say

| Step     | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
| -        | - | -- | -- | -- | - | - | - | - | - |
| T1       | c | a  | b  | e  | d | cc | dd | bb | aa |
| Pressure | 1 | 2  | 2  | 3  | 2 | 3  | 3  | 3 | 2 |


Multi threaded

This PR

| Step     | 0 | 1 | 2 | 3 | 4 |
| -        | - | -- | -- | -- | - |
| T1       | e | d  | bb | aa | b |
| T2       | c | cc | dd | a  | x |
| Pressure | 2 | 3  | 3  | 2  | 2 |


What the docs say

| Step     | 0 | 1 | 2 | 3 | 4 |
| -        | - | -- | -- | -- | - |
| T1       | a | b | d  | bb | aa |
| T2       | c | e | cc | dd | x |
| Pressure | 2 | 2 | 3  | 4  | 2 |

Comment (fjetter) [reply dask/tests/test_order.py:304]: Typically dask.order acts as if we were executing sequentially so what the docs say is not wrong. However, in practice, most will be running in a multi threaded environment. This behavior may, of course, look again very different for more threads...

File: dask/tests/test_order.py [hunk 4]
```diff
@@ -272,27 +465,25 @@ def test_type_comparisions_ok(abcde):
...
-         a
-         |
-      d  b  e
-       \ | /
-         c
+       a
+       |
+    d  b  e
+     \ | /
+       c
```
Comment (fjetter) [inline dask/tests/test_order.py:475]: I believe the argument of the earlier version is wrong. Writing again out what the pressure is over time, yields

Single threaded

This PR

| Step     | 0 | 1 | 2 | 3 | 4 |
| -        | - | -- | -- | -- |-- |
| T1       | c | b  | a | e | d |
| Pressure | 1 | 2  | 2  | 3  | 3 |

Main

| Step     | 0 | 1 | 2 | 3 | 4 |
| -        | - | -- | -- | -- |-- |
| T1       | c | d  | e | b | a |
| Pressure | 1 | 2  | 3  | 3  | 3 |


Multi threaded

This PR

| Step     | 0 | 1 | 2 |
| -        | - | -- | -- |
| T1       | c | b  | a |
| T2       | x | d | e |
| Pressure | 1 | 3  | 3 |


Main

| Step     | 0 | 1 | 2 | 3 |
| -        | - | -- | -- |-- |
| T1       | c | d | b | a |
| T2       | x | e | x | x |
| Pressure | 1 | 3 | 3 | 3 |



Of course, this assumes that all tasks weigh the same. If we assumed that source tasks weight more, this is different but this is not universally true (often but not always).

Favoring the longest critical path, however, benefits execution times in multi threaded environments (fewer steps, again assuming equal weight / runtime)

**Comment by fjetter** [comment]
preliminary A/B tests (not many repeats, i.e. statistics are poor and there were even some failures so please consider this with a grain of salt) are looking promising

Wall time

image

Avg Memory

image

Peak Memory

image



The results look rather good but there are a couple of outliers that need a bit more investigation. 

It's interesting to see that those workflows that benfit from task queuing the most (e.g. anom_mean, basic_sum, vorticity) are almost not affected by this (presumably because their real ordering is mostly fixed by the queuing)

**Comment by mrocklin** [comment]
This is exciting.

File: dask/base.py [hunk 5]
```diff
@@ -718,9 +718,10 @@ def visualize(
 
     https://docs.dask.org/en/latest/optimize.html
     """
-    args, _ = unpack_collections(*args, traverse=traverse)
+    dsk = args[0]
+    # args, _ = unpack_collections(*args, traverse=traverse)
 
-    dsk = dict(collections_to_dsk(args, optimize_graph=optimize_graph))
+    # dsk = dict(collections_to_dsk(args, optimize_graph=optimize_graph))
```
Comment (fjetter) [inline dask/base.py:725]: - [ ] revert before merge

File: dask/order.py [hunk 6]
```diff
@@ -736,10 +684,7 @@ def finish_now_key(x):
 
                         inner_stack_pop = inner_stack.pop
                     for key, vals in dep_pools.items():
-                        if key < item_key:
-                            next_nodes[key].append(vals)
-                        else:
-                            later_nodes[key].append(vals)
+                        next_nodes[key].append(vals)
```
Comment (fjetter) [inline dask/order.py:687]: - [ ] TODO: Considering replicating possible_singles logic here but don't have tests for it

**Comment by fjetter** [comment]
Unfortunately, this still require a bit more work. There is a major regression in the [`test_double_diff`
](https://github.com/coiled/benchmarks/blob/99a8270d62dbe43b5f434403f0456d7e94bf63f8/tests/benchmarks/test_array.py#L219C10-L231)

which is basically the one liner

`diff = a[1:, 1:] - b[:-1, :-1]`

I haven't found out why but I'm still optimistic about all of this.

**Comment by fjetter** [comment]
Ok, I think this is mostly good to go

Cutting off the noise, the most important improvement is the xarray reduction that is also listed in our benchmarks as `test_quadratic_mean` and there appears to be a minor improvement for `map_overlap` as well. Most other _signals_ on the result look like noise (maybe the new algorithm is also slightly faster for these specific cases but I haven't measured)

Wall time this PR
image
Peak mem this PR
image

Zooming out on the Wall Time this looks generally very positive


image

The red excess at the bottom are the tests `test_zarr.py::test_access_slices` which are slicing array components and the entire test is heavily IO dependent. The graph itself is not impacted by ordering since it is 100% embarassingly parallel after optimization. Main also shows a similar, albeit smaller drift.

Looking at peak memory, this becomes a little more interesting. Similar to the wall time, map_overlap and quadratic_mean are the beneficiaries of this change. The `test_double_diff` shows a major (relative) regression at first glance that warrants a closer investigation.

image

# test_double_diff

## This PR 

image

## Main

image

Indeed, the new PR version requires about twice as much memory but we're still very well within a reasonable range.

In fact, the changes here can be traced back to changes I made later in this PR
image


This PR

DoubleDiff_NewOrdering

Main

Screen Recording 2023-10-05 at 09 58 34

If we compare the two version we can see that both are being processed very smoothly. In fact, even with turned off task queuing this behaves very well.

Squinting a little, we can see that we're holding a couple more `rechunk-merge` tasks in memory with this new PR which I strongly suspect is triggered by a change I made to how eagerly _singles_ are being processed (i.e. dangling dependents of a task) so this change is more or less by design. The more eager processing of singles is ultimately what accelerates the `quadratic_mean` so I'm happy for this exchange. The reason why this didn't show up in earlier version of this PR is because I had to change some of the logic to make all tests pass. All things considered, I think this very mild regression is worth it.

File: dask/order.py [hunk 7]
```diff
@@ -460,51 +426,48 @@ def finish_now_key(x):
...
                 deps = set()
                 add_to_inner_stack = True if inner_stack or inner_stacks else False
-                for single, parent in singles_items:
-                    if single in result:
-                        continue
+                singles_keys = set_difference(set(singles), result)
+
+                # NOTE: If this was too slow, LIFO would be a decent
+                # approximation
+                for single in sorted(singles_keys, key=lambda x: partition_keys[x]):
```
Comment (fjetter) [inline dask/order.py:436]: I didn't perform any measurements but empirically, I found that many singles are already part of result unless we're working in very dense all-to-all graphs.

In practice, this means that `singles` should be small here and sorting is feasible. In earlier versions I was simply iterating reversely which also provided really nice results but sorting it is much better.

This is mostly replacing the finish_now logic that I removed above. Cherry on top is that this may now even finish dangling linear branches instead of just single tasks. The downside is that it is no longer _guaranteed_

File: dask/order.py [hunk 8]
```diff
@@ -460,51 +426,48 @@ def finish_now_key(x):
...
-                        add_to_inner_stack
-                        and len(set_difference(dependents[parent], result)) > 1
+                        len(
+                            set_difference(
+                                set_difference(dependents[parent], result),
+                                seen,
+                            )
+                        )
+                        > 1
                     ):
```
Comment (fjetter) [inline dask/order.py:455]: This is what makes the singles processing actually a little more eager than before. Apart from the topological normalization, probably the most impactful change in here.
We're typically only allowing singles to be processed if they are guaranteed to release their parent. this condition now also allows to run a single to run if the parent would be freed after the next tactical goal is reached (i.e. inner_stack is emptied).
This is very close to the idea of later_singles that is already implemented here but there are multiple code paths that create a new inner_stack without triggering a singles compute. I found this logic to be more reliable.

File: dask/order.py [hunk 9]
```diff
@@ -556,30 +518,18 @@ def finish_now_key(x):
...
-                        inner_stack = [dep]
-                        inner_stack_pop = inner_stack.pop
-                        seen_add(dep)
-                        continue
+                if add_to_inner_stack and not inner_stack:
+                    inner_stack = [dep]
+                    inner_stack_pop = inner_stack.pop
+                    seen_add(dep)
+                    continue
+                key = partition_keys[dep]
```
Comment (fjetter) [inline dask/order.py:526]: No functional change, should just be a reduction of code duplication

File: dask/order.py [hunk 10]
```diff
@@ -682,10 +615,12 @@ def finish_now_key(x):
...
                             now_keys.append(key)
-                        elif key < item_key:
-                            next_nodes[key].append(vals)
                         else:
-                            later_nodes[key].append(vals)
+                            psingles = possible_singles[key]
+                            for s in psingles:
+                                singles[s] = item
+                            vals -= psingles
+                            next_nodes[key].append(vals)
```
Comment (fjetter) [inline dask/order.py:623]: This is another big-ish change. Previously, _singles_ would only be considered for nodes with two dependents. This now allows us to detect dependents of nodes that split into many tasks to be treated as singles as well.

File: dask/order.py [hunk 11]
```diff
@@ -819,14 +751,16 @@ def graph_metrics(dependencies, dependents, total_dependencies):
...
+        approximation.
 
-        1
+        0
         |
-        2   1
+        1   0
          \ /
-          4
+          3
```
Comment (fjetter) [inline dask/order.py:763]: TODO: revert

**Review by fjetter — COMMENTED** [review]
I left inline comments to explain all the changes that happened here. This can either guide a reviewer or help a forensic analysis in case something blows up after all

**Comment by mrocklin** [comment]
> Ok, I think this is mostly good to go

This has been fun to read every morning (certainly more interesting and promising than the daily news).  Thanks!

It's a shame we no longer write papers. This would be an interesting topic.

**Comment by mrocklin** [comment]
It seems like the relative nature of the AB tests may bemisleading sometimes.  Might be worth rethinking at some point.

**Comment by fjetter** [comment]
I intend to follow up here with with a smaller non-functional PR that cleans this up a little further like...

- a couple of variable renames
- maybe typing if I feel like it
- Another pass on the doc strings
- if I feel lucky, I'll refactor some duplicated code 🤞 

but I think all of this is better left for a follow up PR

**Comment by d-v-b** [comment]
I have loved reading this, and it's very exciting!

**Comment by mrocklin** [comment]
https://twitter.com/mrocklin/status/1709895827341095184

**Comment by fjetter** [comment]
I intend to merge in about 24hrs unless there is a valid reason not to. If a review comes in after this, I will address any raised issues in a follow-up PR.

Merging early allows this to sit on main (and be tested by dask engineers and our benchmarking suite) for a couple of days before the next release.

**Comment by mrocklin** [comment]
+1

File: dask/order.py [hunk 12]
```diff
@@ -365,35 +365,21 @@ def finish_now_key(x):
                         seen_update(deps)
                     if not singles:
                         continue
+                    # Only process singles once the inner_stack is fully
+                    # resolved, i.e. there is a runnable task This is important
```
Comment (hendrikmakait) [inline dask/order.py:369]:
```suggestion
                    # resolved, i.e., there is a runnable task. This is important
```

**Review by hendrikmakait — APPROVED** [review]
Thanks for adding the comments. They make it ~easier~ feasible to understand what's going on. Generally, the commentary and testing seem reasonable to me. I haven't done an in-depth check of the functional changes, but at least there is nothing obvious going awry. This is good enough for me to merge 👍

File: dask/order.py (hunk 11 above)
Comment (hendrikmakait) [reply dask/order.py:763]: Has this been addressed?

File: dask/order.py [hunk 13]
```diff
@@ -1076,3 +1010,40 @@ def diagnostics(dsk, o=None, dependencies=None):
...
+                else:
+                    new_spec.append(el)
+            elif isinstance(el, list):
+                new_spec.append([_convert_task(e) for e in el])
+        return (_f, *new_spec)
+    else:
+        return task
+
+
+def sanitize_dsk(dsk):
```
[the remark this answers is not in the document]
Comment (fjetter) [reply dask/order.py:1038]: > nit: Adding some type hints to new functions (and order) would be a nice addition. The code is already difficult enough to understand if you know what data types to expect.

I intend to follow up with this (also some variable renaming) but this is definitely out of scope for this PR

File: dask/order.py [hunk 14]
```diff
@@ -213,14 +223,11 @@ def dependencies_key(x):
...
-        return (-len(dependencies[x]), StrComparable(x))
-
+    root_total_dependencies = total_dependencies[list(root_nodes)[0]]
     # Computing this for all keys can sometimes be relatively expensive :(
     partition_keys = {
         key: (
-            (min_dependencies - total_dependencies[key] + 1)
+            (root_total_dependencies - total_dependencies[key] + 1)
             * (total_dependents - min_heights)
         )
```
[the remark this answers is not in the document]
Comment (fjetter) [reply dask/order.py:232]: I don't understand this cost function sufficiently well to say whether this must be strictly positive. I merely discovered cases where zero caused very bad behavior. I think this PR is a better documentation than whatever I can write in the code.

**Comment by fjetter** [comment]
In case anybody is still stumbling over this PR, we recently merged a follow up https://github.com/dask/dask/pull/10660 that is rewriting this entire code section. If you are encountering issues with this, please open a new ticket
